src/libespeak-ng/translate.c

```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+
+	// Do replacement writing 'to' part in buffer
+	while (*pcharto != 0 ) {
+		*ptmpsource = *pcharto;
+		pcharto++;
+		ptmpsource++;
+	}
+
+	// If 'to' part is different than 'from' part, fix remaining part of the buffer.
+	// (Read after 'from' in source, write after 'to' in buffer.
+	// Don't touch last 40 bytes of source, which could contain control data.
+	if (usedfrom != usedto) {
+		psource = *matchstart + usedfrom;
+		while (*psource != 0 && psource < &source[N_TR_SOURCE]
+				&& ptmpsource < &tmpsource[N_TR_SOURCE]) {
+			*ptmpsource = *psource;
+			ptmpsource++;
+			psource++;
+		}
 	}
-	if (c == 0) return 0;
-
-	if ((replace_chars = tr->langopts.replace_chars) == NULL)
-		return c;
 
-	// there is a list of character codes to be substituted with alternative codes
-
-	if (iswupper(c_lower = c)) {
-		c_lower = towlower2(c, tr);
-		upper_case = 1;
+	// Write temporary buffer back to source
+	psource = *matchstart;
+	ptmpsource = &tmpsource;
+	while (*ptmpsource != 0 && ptmpsource < &tmpsource[N_TR_SOURCE]
+			&& psource < &source[N_TR_SOURCE]) {
+		*psource = *ptmpsource;
+		ptmpsource++;
+		psource++;
 	}
 
-	new_c = 0;
-	for (ix = 0; (word = replace_chars[ix]) != 0; ix += 2) {
-		if (c_lower == (word & 0xffff)) {
-			if ((word >> 16) == 0) {
-				new_c = replace_chars[ix+1];
-				break;
-			}
-			if ((word >> 16) == (unsigned int)towlower2(next_in, tr)) {
-				new_c = replace_chars[ix+1];
-				ignore_next = true;
-				break;
-			}
-		}
+	// if requested, print trace
+	if (option_phonemes & espeakPHONEMES_TRACE) {
+		fprintf(f_trans, "Replace   %s > %s\n", charfrom, charto);
 	}
+}
 
-	if (new_c == 0)
-		return c; // no substitution
+static void TranslateChars(Translator *tr, char *source)
+{
+	// Replace characters using Translator.replace_chars in passed source buffer
+	int *replace_chars = tr->langopts.replace_chars;
+	char *mysource = source;
+	char *startmatch;
+	int *from;
+	int *to;
+	bool upper_case = false;
+
+	if ((from = tr->langopts.replace_chars) == NULL)
+		return;
 
-	if (new_c & 0xffe00000) {
-		// there is a second character to be inserted
-		// don't convert the case of the second character unless the next letter is also upper case
-		c2 = new_c >> 16;
-		if (upper_case && iswupper(next_in))
-			c2 = ucd_toupper(c2);
-		*insert = c2;
-		new_c &= 0xffff;
+	char *curchar = source; // pointer to the source clause
+	int curint;             // current UTF-16 character
+	int used;               // bytes used in buffer
+	int *curfrom;           // pointer to current 'from'
+	bool next_step = false; // signal to jumping to next 'from' part in dictionary
+	while (GetNextTranslateChars(&from, &to, next_step)) {
+		next_step = true;
+		curfrom = from;
+		curchar = source;
+		startmatch = 0;
+		curint = 0;
+		do {
+			used = utf8_in(&curint, curchar);
+			if (curint < ' ') {        // skip control characters
+				curchar += used;
+				continue;
+			}
+			if ((upper_case = iswupper(curint)) == true) { // remember, if uppercase
```

Comment:
[Minor] A single-statement if statement does not need braces. This is to keep a consistent coding style with the rest of the codebase.